platform/api/oci: add oci client

arithx opened the pull request:
Adds the http.Client which will automatically sign requests.

cc @bgilbert

bgilbert commented on platform/api/oci/client_test.go:
```diff
@@ -0,0 +1,182 @@
+	profile := GetDefaultAuthProfile(t)
+	profile.KeyID = "ocid1.tenancy.oc1..aaaaaaaaba3pv6wkcr4jqae5f15p2b2m2yt2j6rx32uzr4h25vqstifsfdsq/ocid1.user.oc1..aaaaaaaat5nvwcna5j6aqzjcaty5eqbb6qt2jvpkanghtgdaqedqw3rynjq/20:3b:97:13:55:1c:5b:0d:d3:37:d8:50:4e:c5:3a:34"
+
+	client := NewHTTPClient(context.TODO(), &profile)
```
If it's possible without a lot of work, it might make sense to use a base RoundTripper which doesn't actually make a network request.

arithx commented:
Pushed updates addressing our OOB discussions.

- Updated `auth/oci.go` to use `go-ini/ini` based on the `bmcs` tool's config.
- General refactoring to accomodate the `auth/oci.go` change
- Updated `client_test.go` to use a MockTransport rather than sending the actual HTTP requests
- Updated `go-ini` vendor to `1.28.2`

bgilbert commented on auth/oci.go:
```diff
@@ -0,0 +1,99 @@
+	"github.com/go-ini/ini"
+)
+
+const OCIConfigPath = ".oraclebmc/mantle"
```
If this doesn't need to be exported, let's keep it internal to the package to avoid temptation.

bgilbert commented on auth/oci.go:
```diff
@@ -0,0 +1,99 @@
+// ReadOCIConfig decodes an OCI config file, for more information:
+// https://docs.us-phoenix-1.oraclecloud.com/Content/API/Concepts/sdkconfig.htm
+//
+// If path is empty, $HOME/.oraclebmc/mantle is read.
```
The thinking was that we'd read the existing `config` file for keys in the official schema, and `mantle` for additional keys like `compartment`.  I guess that means the profile struct will need to have keys from both files.

bgilbert commented on auth/oci.go:
```diff
@@ -0,0 +1,99 @@
+			return nil, err
+		}
+
+		profiles[strings.ToLower(section.Name())] = p
```
Is the file case-insensitive?  I assume from the use of the section `DEFAULT` that it isn't.

arithx replied:
The file is being read via `InsensitiveLoad` which ignores casing for sections & keys.

Lowercasing the profile name was to keep it in similar to other APIs where they default the profile name to `default` if it's not provided (https://github.com/coreos/mantle/blob/master/platform/api/packet/api.go#L123)

bgilbert commented on auth/oci.go:
```diff
@@ -0,0 +1,99 @@
+	} else {
+		return priv, nil
+	}
+}
```
Sadly, we should probably check that the key matches the fingerprint, since otherwise there will be confusion later.

arithx replied:
Working on adding this back in, the tests in `platform/api/oci/client_test.go` will have to be updated as a result because they're using the example data provided by Oracle in the docs and the provided RSA key does not match the provided fingerprint they use in the KeyID.

bgilbert commented on platform/api/oci/client_test.go:
```diff
@@ -0,0 +1,208 @@
+7U1yQXnTAEFYM560yJlzUpOb1V4cScGd365tiSMvxLOvTA==
+-----END RSA PRIVATE KEY-----`
+	rsaPath := filepath.Join(tempDir, "tempRSA.pem")
+	file, err := os.Create(rsaPath)
```
Leaked.

arithx replied:
Adding a a `defer file.Close()` call after the error check.

The cleanup of the file is taken care of via the `defer os.RemoveAll(tempDir)` calls in the `TestAuthOCIHelperMethods` & `TestHTTPClient` methods.

bgilbert replied:
Right, the file handle is what I meant.

bgilbert commented on platform/api/oci/client_test.go:
```diff
@@ -0,0 +1,208 @@
+	return response, nil
+}
+
+func GetDefaultAuthProfile(t *testing.T, tempDir string) auth.OCIProfile {
```
Should be private?

arithx commented:
Pushed changes
